core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java

```diff
@@ -276,6 +276,9 @@ public VolcanoPlanner(RelOptCostFactory costFactory, //
     super(costFactory == null ? VolcanoCost.FACTORY : costFactory, //
         externalContext);
     this.zeroCost = this.costFactory.makeZeroCost();
+    // If LOGGER is debug enabled, enable provenance information to be captured
+    this.provenanceMap = LOGGER.isDebugEnabled() ? new HashMap<>()
+        : Util.blackholeMap();
```

Comment:
Do we need a black hole map? Can't we just use this?
```
this.provenanceMap = new HashMap<RelNode, Provenance>() {
    @Override public Provenance put(final RelNode key, final Provenance value) {
      return LOGGER.isDebugEnabled() ? super.put(key, value) : null;
    }
  };
```

Reply:
- all the put variants need to be overriden, so at that point it's probably simpler to subclass AbstractMap (which is what blackhole map does).
- blackhole map is explicit about its behavior (vs an anonymous inner class). When debugging, you know what provenanceMap behavior is just by looking at the class name.
- logger level can be changing during planning externally, so better to only check once the level vs doing it each time `put` is used (and possibly ending up with a partial map)

Reply:
-1 for black hole map, the name is really confusing.

Reply:
Naming is very subjective. It is would be helpful if you provide an alternative suggestion if you're unhappy with a name. What do people think about NullMap, DevNullMap, and ForgetfulMap?

Reply:
I agree with @jacques-n that naming is subjective, and I would also recommend to explain why it is confusing and if you have suggestions. Blackhole is used regularly in programming for methods/classes which can be used to write data, but you cannot read it back (similar to a black hole in astronomy). My inspiration came from JMH framework and their `Blackhole` class: https://hg.openjdk.java.net/code-tools/jmh/file/1c11c886e0c8/jmh-core/src/main/java/org/openjdk/jmh/infra/Blackhole.java#l147. I'd be okay with `DevNullMap` (assuming everybody is familiar with `/dev/null`...)

I'd like also to avoid too much bikeshedding on the name. If name is the only comment left, and there's no strong consensus for or against a name, I'll plan to merge this change in 1 or 2 days...